cpp/include/raft/interruptible.hpp

```diff
@@ -0,0 +1,246 @@
+#include <raft/cudart_utils.h>
+#include <raft/error.hpp>
+#include <rmm/cuda_stream_view.hpp>
+#include <thread>
+#include <unordered_map>
+
+namespace raft {
+
+/**
+ * @brief Exception thrown during `interruptible::synchronize` call when it detects a request
+ * to cancel the work performed in this CPU thread.
+ */
+struct interrupted_exception : public raft::exception {
+  explicit interrupted_exception(char const* const message) : raft::exception(message) {}
+  explicit interrupted_exception(std::string const& message) : raft::exception(message) {}
+};
+
+class interruptible {
+ public:
+  /**
+   * @brief Synchronize the CUDA stream, subject to being interrupted by `interruptible::cancel`
+   * called on this CPU thread.
+   *
+   * @param [in] stream a CUDA stream.
+   *
+   * @throw raft::interrupted_exception if interruptible::cancel() was called on the current CPU
+   * thread id before the currently captured work has been finished.
+   * @throw raft::cuda_error if another CUDA error happens.
+   */
+  static inline void synchronize(rmm::cuda_stream_view stream)
+  {
+    get_token()->synchronize_impl(stream);
+  }
+
+  /**
+   * @brief Check the thread state, whether the thread is interrupted by `interruptible::cancel`.
+   *
+   * This is a cancellation point for an interruptible thread. It's called in the internals of
+   * `interruptible::synchronize` in a loop. If two synchronize calls are far apart, it's
+   * recommended to call `interruptible::yield()` in between to make sure the thread does not become
+   * unresponsive for too long.
+   *
+   * Both `yield` and `yield_no_throw` reset the state to non-cancelled after execution.
+   *
+   * @throw raft::interrupted_exception if interruptible::cancel() was called on the current CPU
+   * thread.
+   */
+  static inline void yield() { get_token()->yield_impl(); }
+
+  /**
+   * @brief Check the thread state, whether the thread is interrupted by `interruptible::cancel`.
+   *
+   * Same as `interruptible::yield`, but does not throw an exception if the thread is cancelled.
+   *
+   * Both `yield` and `yield_no_throw` reset the state to non-cancelled after execution.
+   *
+   * @return whether interruptible::cancel() was called on the current CPU thread.
+   */
+  static inline auto yield_no_throw() -> bool { return get_token()->yield_no_throw_impl(); }
+
+  /**
+   * @brief Get a cancellation token for this CPU thread.
+   *
+   * @return an object that can be used to cancel the GPU work waited on this CPU thread.
+   */
+  static inline auto get_token() -> std::shared_ptr<interruptible>
+  {
+    static thread_local std::shared_ptr<interruptible> s(get_token(std::this_thread::get_id()));
+    return s;
+  }
+
+  /**
+   * @brief Get a cancellation token for a CPU thread given by its id.
+   *
+   * @return an object that can be used to cancel the GPU work waited on the given CPU thread.
+   */
+  static auto get_token(std::thread::id thread_id) -> std::shared_ptr<interruptible>
```

Comment:
Thread IDs are frequently recycled. I haven't looked at this close enough to say whether this could be a problem or not, but you should think through to make sure that this code won't break if over the course of the program two distinct threads call `get_token` with the same thread id.

Reply:
I think what I'm more worried about is if a thread ID can be recycled before all of the thread's `thread_local` objects have been destroyed. e.g., if an ID is recycled before the destroyed thread's `thread_local shared_ptr<interruptible>` has been destroyed, then that would probably lead to all sorts of bad things.

Reply:
Ok, the new `thread_seq_id_` should be able to take care of both of these cases.